pkg/lockfile/apk-installed.go

```diff
@@ -58,21 +58,26 @@ func parseApkPackageGroup(group []string, pathToLockfile string) PackageDetails
 			pkgPrintName = unknownPkgName
 		}
 
-		_, _ = fmt.Fprintf(
-			os.Stderr,
-			"warning: malformed APK installed file. Found no version number in record. Package %s. File: %s\n",
+		diag.Warn(fmt.Sprintf(
+			"warning: malformed APK installed file. Found no version number in record. Package %s. File: %s",
 			pkgPrintName,
 			pathToLockfile,
-		)
+		))
 	}
 
 	return pkg
 }
 
 func ParseApkInstalled(pathToLockfile string) ([]PackageDetails, error) {
+	return parseFileAndPrintDiag(pathToLockfile, ParseApkInstalledWithDiagnostics)
+}
+
+func ParseApkInstalledWithDiagnostics(pathToLockfile string) ([]PackageDetails, Diagnostics, error) {
```

Comment:
FWIW, thinking about a separate `Diagnostics` return value: in HCL/Terraform-land, the [diagnostics](https://github.com/hashicorp/terraform/blob/3d1a58d5b54bdcde038a6f06150e566f94e94357/internal/tfdiags/diagnostics.go#L15-L22) are returned _like an error_, not a separate thing. Diagnostics [carry a severity](https://github.com/hashicorp/terraform/blob/3d1a58d5b54bdcde038a6f06150e566f94e94357/internal/tfdiags/diagnostic.go#LL10) that determines if they're either a [warning](https://github.com/hashicorp/terraform/blob/3d1a58d5b54bdcde038a6f06150e566f94e94357/internal/tfdiags/diagnostic.go#L33) or an [error](https://github.com/hashicorp/terraform/blob/3d1a58d5b54bdcde038a6f06150e566f94e94357/internal/tfdiags/diagnostic.go#L32). 

The convention becomes to use the [`diags.HasErrors()`](https://github.com/search?q=repo%3Ahashicorp/terraform%20HasErrors&type=code) method to check if the returned diagnostics have any `error`s that need to be dealt with; otherwise you know the rest are just warnings that might be useful to log, or show the user, but not halt further execution.

While this is a _minor_ detail, I think the pattern is useful in these situations.

```go
pkgDetails, diags := ParseApkInstalledWithDiagnostics(pathToLockfile)
if diags.HasErrors() {
  // add info and return an `error` if appropriate
  return fmt.Errorf("failed to parse APK file %q in %q: %w", pathToLockfile, containerImage, diags.Err())
}
```

Either way, I want diagnostics! But, I figured I'd at least bring this up if you want to discuss it further 😄

Reply:
hmm that's interesting - I think it makes sense to consider because there's no real downside so long as `diags.Err()` exposes the underlying error? (actually, the main one I can think of is that IDEs like Goland would not tell you that you need to check if there was an error, which I'm pretty sure it does based on if the return is an `error`).

Though I'm not sure if there's an advantage to returning an object vs a slice?